File: GroupICAT/icatb/tests/complex_ica/test_phase_correct.m
function test_phase_correct()
addpath(genpath(fullfile(fileparts(mfilename('fullpath')), '..', '..', 'icatb_analysis_functions', 'icatb_algorithms', 'complex_ica')));
rng(5);
N = 4; V = 3000; M = 20;
% real-axis-aligned sources (elongated along real axis), then inject a known rotation
Sbase = (randn(N,V)) + 1i*(0.05*randn(N,V));      % energy mostly real
A0 = randn(M,N) + 1i*randn(M,N);
theta_true = [0.7; -1.1; 0.3; 2.0];
Srot = Sbase .* exp(1i*repmat(theta_true, 1, V));
Arot = A0 .* exp(-1i*repmat(theta_true.', M, 1)); % keep A*S invariant
mask = true(V,1);

% PROPERTY 1: X = A*S preserved by the whole correction
X_before = Arot*Srot;
[Sc, Ac, theta] = icatb_complex_phase_correct(Srot, Arot, mask);
assert(norm(X_before - Ac*Sc, 'fro') / norm(X_before, 'fro') < 1e-10, 'A*S must be preserved');

% PROPERTY 2: corrected sources are real-axis aligned (imag energy minimized)
imag_frac = sum(imag(Sc).^2, 2) ./ sum(abs(Sc).^2, 2);
assert(all(imag_frac < 0.05), 'corrected sources must concentrate energy on the real axis');

% PROPERTY 3: recovered rotation cancels the injected one (up to pi sign)
resid = mod(angle(sum(Sc.^2, 2)), pi);
assert(all(min(resid, pi - resid) < 1e-3), 'residual major-axis angle ~ 0 mod pi');
disp('PASS test_phase_correct');
end

File: python/tests/test_phase_correct.py
import numpy as np

from complex_gift.phase_correct import align_to_reference, correct_phase


def _aligned_sources(rng, N=4, V=3000):
    """Sources whose energy already lies along the real axis."""
    return rng.standard_normal((N, V)) + 1j * 0.05 * rng.standard_normal((N, V))


def test_reconstruction_is_preserved():
    """The whole point: rotating S by e^{i0} and A by e^{-i0} must leave A @ S alone."""
    rng = np.random.default_rng(5)
    N, V, M = 4, 3000, 20
    S0 = _aligned_sources(rng, N, V)
    A0 = rng.standard_normal((M, N)) + 1j * rng.standard_normal((M, N))
    theta_true = np.array([0.7, -1.1, 0.3, 2.0])
    S = S0 * np.exp(1j * theta_true)[:, None]
    A = A0 * np.exp(-1j * theta_true)[None, :]

    X_before = A @ S
    Sc, Ac, theta = correct_phase(S, A)
    rel = np.linalg.norm(X_before - Ac @ Sc) / np.linalg.norm(X_before)
    assert rel < 1e-10


def test_injected_rotation_is_removed():
    """A known rotation is undone: corrected sources concentrate on the real axis."""
    rng = np.random.default_rng(5)
    N, V, M = 4, 3000, 20
    S0 = _aligned_sources(rng, N, V)
    A0 = rng.standard_normal((M, N)) + 1j * rng.standard_normal((M, N))
    theta_true = np.array([0.7, -1.1, 0.3, 2.0])
    S = S0 * np.exp(1j * theta_true)[:, None]
    A = A0 * np.exp(-1j * theta_true)[None, :]

    Sc, Ac, theta = correct_phase(S, A)
    imag_frac = (Sc.imag**2).sum(axis=1) / (np.abs(Sc) ** 2).sum(axis=1)
    assert np.all(imag_frac < 0.05)
    # residual major-axis angle is ~0 mod pi
    resid = np.mod(np.angle((Sc**2).sum(axis=1)), np.pi)
    assert np.all(np.minimum(resid, np.pi - resid) < 1e-3)


def test_mask_restricts_the_estimate():
    """theta is estimated only from masked (high-quality) voxels: garbage outside the
    mask must not move the answer."""
    rng = np.random.default_rng(6)
    N, V, M = 3, 2000, 10
    S = _aligned_sources(rng, N, V)
    A = rng.standard_normal((M, N)) + 1j * rng.standard_normal((M, N))
    mask = np.zeros(V, dtype=bool)
    mask[: V // 2] = True
    S_dirty = S.copy()
    S_dirty[:, ~mask] *= 50.0 * np.exp(1j * 1.3)      # wild values outside the mask

    _, _, theta_clean = correct_phase(S.copy(), A.copy(), mask=mask)
    _, _, theta_dirty = correct_phase(S_dirty, A.copy(), mask=mask)
    assert np.allclose(theta_clean, theta_dirty, atol=1e-9)


def test_group_alignment_collapses_known_subject_rotations():
    rng = np.random.default_rng(7)
    N, V = 3, 1500
    S_ref = _aligned_sources(rng, N, V)
    rot = np.array([0.9, -0.4, 2.2])
    S_subj = S_ref * np.exp(1j * rot)[:, None]
    S_al, theta = align_to_reference(S_subj, S_ref)
    assert np.allclose(np.abs(theta), np.abs(rot), atol=1e-6)
    # aligned sources now agree with the reference up to a positive real scale
    for k in range(N):
        c = np.vdot(S_ref[k], S_al[k]) / np.vdot(S_ref[k], S_ref[k])
        assert abs(c.imag) < 1e-6
